### DISCUSSION TITLE
USB MIDI 2.0 with fallback to USB MIDI 1.0 Drivers Available

### DISCUSSION THREAD

**Discussion by AmeNote-Michael** [hathach/tinyusb#2457] [General]
In conjunction with the MIDI Association and several members, AmeNote has created and released a tinyUSB device driver for USB MIDI 2.0.
[tusb_ump](https://github.com/midi2-dev/tusb_ump)

Currently there are no examples included, however you are welcome to view and use source for our MIDI 2.0 prototyping tool to see how tusb_ump is used.

[AmeNote_ProtoZOA](https://github.com/midi2-dev/AmeNote_Protozoa)

Both the tusb_ump and AmeNote_ProtoZOA code has been tested on Raspberry Pico devices extensively with limited testing of tusb_ump on other platforms.

We welcome your inputs and contributions to the tusb_ump library. It is our hope to eventually supplement or replace the existing MIDI driver provided with tinyUSB.

If you are interested in MIDI 2.0 in general, please check out our public available tools and libraries.
[MIDI2.DEV](https://midi2.dev/)

Also the [MIDI Association](https://www.midi.org/) has many great resources.

**Comment by hathach** [comment]
thank you very much for your work, would you mind if I integrate your ump driver as part of tinyusb built-in driver for MIDI. To remain backward compatible with existing MIDI 1.0 driver, we could introduce a config macro such as `CFG_TUD_MIDI_USE_UMP` or `CFG_TUD_MIDI_V2` to explicitly enable midi v2 support using UMP driver. That would make it easier for user to quickly get MIDI v2 driver.

I am off for TET holidays for a couple of weeks. Will try to test this out afterwards.

**Reply by AmeNote-Michael** [reply]
It would be our pleasure to contribute to tinyUSB, it was our eventual intent anyways. Perhaps, time permitting, I can create a simple general example for the UMP driver in next two weeks.
